In [1]:
# Load libraries

# Data Manipulation
import pandas as pd
# Custom transformers
from custom_transformers import Cleaner, StopWordsRemover, Lemmatizer
# TF-IDF Vectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

# Download the Spanish model
# import stanza
# stanza.download("es")

In [2]:
# Load the dataset
dataset = pd.read_csv("./../datasets/dataset_labeled_v2.csv")
# dataset = dataset[293:298].copy()
# Load stop words
with open("stop_words.txt", "r") as f:
    stop_words = f.read().splitlines()

print(stop_words)
display(dataset.head())
print(dataset.info())

['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '_', 'a', 'actualmente', 'acuerdo', 'adelante', 'ademas', 'además', 'adrede', 'afirmó', 'agregó', 'ahi', 'ahora', 'ahÃ\xad', 'al', 'algo', 'alguna', 'algunas', 'alguno', 'algunos', 'algún', 'alli', 'allÃ\xad', 'alrededor', 'ambos', 'ampleamos', 'antano', 'antaño', 'ante', 'anterior', 'antes', 'apenas', 'aproximadamente', 'aquel', 'aquella', 'aquellas', 'aquello', 'aquellos', 'aqui', 'aquél', 'aquélla', 'aquéllas', 'aquéllos', 'aquÃ\xad', 'arriba', 'arribaabajo', 'aseguró', 'asi', 'asÃ\xad', 'atras', 'aun', 'aunque', 'ayer', 'añadió', 'aún', 'b', 'bajo', 'bastante', 'bien', 'breve', 'buen', 'buena', 'buenas', 'bueno', 'buenos', 'c', 'cada', 'casi', 'cerca', 'cierta', 'ciertas', 'cierto', 'ciertos', 'cinco', 'claro', 'comentó', 'como', 'con', 'conmigo', 'conocer', 'conseguimos', 'conseguir', 'considera', 'consideró', 'consigo', 'consigue', 'consiguen', 'consigues', 'contigo', 'contra', 'cosas', 'creo', 'cual', 'cuales', 'c

id                       date  \
0  1476701687070998535  2021-12-30 23:48:12+00:00   
1  1476682498826452992  2021-12-30 22:31:57+00:00   
2  1476675906903191552  2021-12-30 22:05:46+00:00   
3  1476672423076499456  2021-12-30 21:51:55+00:00   
4  1476669177104588806  2021-12-30 21:39:01+00:00   

                                                text  \
0  #ATENCIÓN ⚠️  Dos presuntos delincuentes fuero...   
1  Y si los hackean @PoliciaEcuador y me terminan...   
2  @EPattoA @PoliciaEcuador Les encanta quedar as...   
3  🙌📢 CONSULTA CONECTADOS | El asalto del año fue...   
4  @EPattoA @PoliciaEcuador 14 años robando Que q...   

                                                 url         location label  
0  https://twitter.com/noticiasru593/status/14767...  ['Samborondón']  robo  
1  https://twitter.com/lujacome/status/1476682498...               []  robo  
2  https://twitter.com/Rengel_A_G/status/14766759...               []  robo  
3  https://twitter.com/conectados_ec/status/14766...     ['Santiago']  robo  
4  https://twitter.com/cesar197449/status/1476669...               []  robo

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 598 entries, 0 to 597
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        598 non-null    int64 
 1   date      598 non-null    object
 2   text      598 non-null    object
 3   url       598 non-null    object
 4   location  598 non-null    object
 5   label     598 non-null    object
dtypes: int64(1), object(5)
memory usage: 28.2+ KB
None


In [3]:
# pipeline = Pipeline(steps=[
#     ("cleaner", Cleaner()),
#     ("stop_words_remover", StopWordsRemover(stop_words=stop_words)),
#     ("lemmatizer", Lemmatizer()),
#     ("tfidf", TfidfVectorizer()),
# ])

df_clean = Cleaner().transform(dataset['text'])
print('Cleaner', df_clean.iloc[293])
df_sw = StopWordsRemover(stop_words=stop_words).transform(df_clean)
print('StopWordsRemover', df_sw.iloc[293])
df_lem = Lemmatizer().transform(df_sw)
print('Lemmatizer', df_lem.iloc[293])
tfidf_vectorizer = TfidfVectorizer()
df_tfidf = tfidf_vectorizer.fit_transform(df_lem)

Cleaner quevedo  hace pocos minutos se registró un caso de raptó varios sujetos secuestraron al dueño de un local comercial  litardo  ubicado en la parroquia  san_camilo     y el  plan fénix
StopWordsRemover quevedo minutos registró caso raptó sujetos secuestraron dueño local comercial litardo ubicado parroquia san_camilo plan fénix


d:\GitHub\tfm\.venv\lib\site-packages\spacy\language.py:1037: UserWarning: Due to multiword token expansion or an alignment issue, the original text has been replaced by space-separated expanded tokens.
  doc = self._ensure_doc(text)
d:\GitHub\tfm\.venv\lib\site-packages\spacy\language.py:1037: UserWarning: Can't set named entities because of multi-word token expansion or because the character offsets don't map to valid tokens produced by the Stanza tokenizer:
Words: ['robaron', 'motorizados', 'parque', 'inglés', 'a', 'ev', 'calle', 'simón', 'cárdenas', 'joaquín', 'soto', 'diciembre', 'dónde', 'denuncia', 'parque', 'inglés']
Entities: [('av calle simón', 'LOC', 34, 48)]
  doc = self._ensure_doc(text)
d:\GitHub\tfm\.venv\lib\site-packages\spacy\language.py:1037: UserWarning: Can't set named entities because of multi-word token expansion or because the character offsets don't map to valid tokens produced by the Stanza tokenizer:
Words: ['ojos', 'águila', 'capta', 'actividad', 'denuncia',

Lemmatizer quevedo minuto registrar caso raptar sujeto secuestrar dueño local comercial litardo ubicado parroquia san_camilo plan fénix


In [46]:
# print(tfidf_vectorizer.vocabulary_)
# print(tfidf_vectorizer.get_feature_names_out())
# print(df_tfidf.toarray())

df_results = pd.DataFrame(df_tfidf.toarray(), columns=tfidf_vectorizer.get_feature_names_out())
df_results[293:294][['quevedo', 'minuto', 'registrar', 'caso', 'raptar', 'sujeto', 'secuestrar', 'dueño', 'local', 'comercial', 'litardo', 'ubicado', 'parroquia', 'san', 'camilo', 'plan', 'fénix']].melt()

variable     value
0      quevedo  0.234918
1       minuto  0.261482
2    registrar  0.271567
3         caso  0.194523
4       raptar  0.302892
5       sujeto  0.221916
6   secuestrar  0.183624
7        dueño  0.261482
8        local  0.234918
9    comercial  0.240241
10     litardo  0.271567
11     ubicado  0.271567
12   parroquia  0.271567
13         san  0.000000
14      camilo  0.000000
15        plan  0.211832
16       fénix  0.225849

quevedo => 0.234918, minuto => 0.261482, registrar => 0.271567, caso => 0.194523, raptar => 0.302892, sujeto => 0.221916, secuestrar => 0.183624, dueño => 0.261482, local => 0.234918, comercial => 0.240241, litardo => 0.271567, ubicado => 0.271567, parroquia => 0.271567, san => 0.000000, camilo => 0.000000, plan => 0.211832, fénix => 0.225849